### PR TITLE
Activate the ``duplicate-code`` checker

### PR DESCRIPTION
**Pull request by Pierre-Sassoulas**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description

This activate the duplicate checkers with a high threshold that we could progressively lower later. I think it detects a false positive with useless suppression / duplicate code when disabling using the scope of the class.

### CODE REVIEW

File: pylint/testutils/unittest_linter.py [hunk 1]
```diff
@@ -15,7 +15,7 @@
 class UnittestLinter:
     """A fake linter class to capture checker messages."""
 
-    # pylint: disable=unused-argument
+    # pylint: disable=unused-argument,duplicate-code
```
Comment (Pierre-Sassoulas) [inline pylint/testutils/unittest_linter.py:18]: Here the whole ``add_message`` function is duplicated from PyLinter so the disable is required but we have a useless suppression warning

**Comment by Pierre-Sassoulas** [comment]
@DanielNoord I think we have a regression in the parsing of option see failed tests 
```
 E   SystemExit: 2
------------------------------- Captured stderr --------------------------------
usage: pylint [-h] [--max-line-length <int>] [--ignore-long-lines <regexp>]
              [--single-line-if-stmt <y or n>]
              [--single-line-class-stmt <y or n>] [--max-module-lines <int>]
              [--indent-string <string>] [--indent-after-paren <int>]
              [--expected-line-ending-format <empty or LF or CRLF>]
              [--max-args <int>] [--max-locals <int>] [--max-returns <int>]
              [--max-branches <int>] [--max-statements <int>]
              [--max-parents <num>]
              [--ignored-parents <comma separated list of class names>]
              [--max-attributes <num>] [--min-public-methods <num>]
              [--max-public-methods <num>] [--max-bool-expr <num>]
              [--exclude-too-few-public-methods <pattern>[,<pattern>...]]
              [--logging-modules <comma separated list>]
              [--logging-format-style <old %) or new ({>]
              [--overgeneral-exceptions <comma-separated class names>]
              [--defining-attr-methods <method names>]
              [--valid-classmethod-first-arg <argument names>]
              [--valid-metaclass-classmethod-first-arg <argument names>]
              [--exclude-protected <protected access exclusions>]
              [--check-protected-access-in-special-methods <y or n>]
              [--deprecated-modules <modules>]
              [--preferred-modules <module:preferred-module>]
              [--import-graph <file.gv>] [--ext-import-graph <file.gv>]
              [--int-import-graph <file.gv>]
              [--known-standard-library <modules>]
              [--known-third-party <modules>]
              [--allow-any-import-level <modules>]
              [--analyse-fallback-blocks <y or n>]
              [--allow-wildcard-with-all <y or n>]
              [--no-docstring-rgx <regexp>] [--docstring-min-length <int>]
pylint: error: ambiguous option: --ignore could match --ignore-long-lines, --ignored-parents
```

**Comment by DanielNoord** [comment]
> @DanielNoord I think we have a regression in the parsing of option see failed tests
> 
> ```
>  E   SystemExit: 2
> ------------------------------- Captured stderr --------------------------------
> usage: pylint [-h] [--max-line-length <int>] [--ignore-long-lines <regexp>]
>               [--single-line-if-stmt <y or n>]
>               [--single-line-class-stmt <y or n>] [--max-module-lines <int>]
>               [--indent-string <string>] [--indent-after-paren <int>]
>               [--expected-line-ending-format <empty or LF or CRLF>]
>               [--max-args <int>] [--max-locals <int>] [--max-returns <int>]
>               [--max-branches <int>] [--max-statements <int>]
>               [--max-parents <num>]
>               [--ignored-parents <comma separated list of class names>]
>               [--max-attributes <num>] [--min-public-methods <num>]
>               [--max-public-methods <num>] [--max-bool-expr <num>]
>               [--exclude-too-few-public-methods <pattern>[,<pattern>...]]
>               [--logging-modules <comma separated list>]
>               [--logging-format-style <old %) or new ({>]
>               [--overgeneral-exceptions <comma-separated class names>]
>               [--defining-attr-methods <method names>]
>               [--valid-classmethod-first-arg <argument names>]
>               [--valid-metaclass-classmethod-first-arg <argument names>]
>               [--exclude-protected <protected access exclusions>]
>               [--check-protected-access-in-special-methods <y or n>]
>               [--deprecated-modules <modules>]
>               [--preferred-modules <module:preferred-module>]
>               [--import-graph <file.gv>] [--ext-import-graph <file.gv>]
>               [--int-import-graph <file.gv>]
>               [--known-standard-library <modules>]
>               [--known-third-party <modules>]
>               [--allow-any-import-level <modules>]
>               [--analyse-fallback-blocks <y or n>]
>               [--allow-wildcard-with-all <y or n>]
>               [--no-docstring-rgx <regexp>] [--docstring-min-length <int>]
> pylint: error: ambiguous option: --ignore could match --ignore-long-lines, --ignored-parents
> ```

Yeah. https://github.com/PyCQA/pylint/pull/6120 needs to land. I forgot it had a fix in it. Sorry about this!

File: pylintrc [hunk 2]
```diff
@@ -163,10 +163,10 @@ ignore-comments=yes
 ignore-docstrings=yes
 
 # Ignore imports when computing similarities.
-ignore-imports=no
+ignore-imports=yes
```
[the remark this answers is not in the document]
Comment (Pierre-Sassoulas) [reply pylintrc:166]: Yes I think so too.

[the remark this answers is not in the document]
Comment (DanielNoord) [reply pylintrc:166]: https://github.com/PyCQA/pylint/pull/6500

[the remark this answers is not in the document]
Comment (DanielNoord) [reply pylintrc:166]: https://github.com/PyCQA/pylint/pull/6500

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas This should be ready for review!

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas This should be ready for review!

**Comment by DanielNoord** [comment]
I'm going to draft this and extract some of the changes in here.

For some reason `pre-commit` locally and `pre-commit` in CI has different output for `pre-commit run pylint --all-files`. We'll need to investigate this before we can merge this.

File: tests/test_similar.py [hunk 3]
```diff
@@ -91,6 +92,7 @@ def test_duplicate_code_raw_strings_disable_file(self) -> None:
                 "--enable=duplicate-code",
                 "--ignore-imports=no",
                 "--ignore-signatures=no",
+                "--min-similarity-lines=4",
```
Comment (Pierre-Sassoulas) [inline tests/test_similar.py:95]: At this point we could use a pylintrc for similar tests instead of duplicating the option in all calls ?

Comment (DanielNoord) [reply tests/test_similar.py:95]: Yeah I thought about doing something like that, but then I thought: it's actually better to be explicit about this. Kind of strange how us changing our `pylintrc` is breaking tests. This makes it much clearer what we are actually testing.

Comment (Pierre-Sassoulas) [reply tests/test_similar.py:95]: > Kind of strange how us changing our pylintrc is breaking tests.

I agree that's why I did #6459, it should not happen again (I'm surprised it does as I tested this by raising an exception in init-hook). What I mean is create a specific pylintrc for the similar tests. But it's a little less explicit than calling all the options in the tests itself, you do have to go check the content of the pylintrc.

**Comment by Pierre-Sassoulas** [comment]
Moving to 2.15 !

File: tests/test_similar.py (hunk 3 above)
Comment (DanielNoord) [reply tests/test_similar.py:95]: Yeah, I think this is fine for now. I promise to revisit when our issue count is less than 10 😄

**Comment by DanielNoord** [comment]
> Moving to 2.15 !

Almost done. Please see https://github.com/PyCQA/pylint/pull/6504 😄 🎉

File: tests/test_similar.py (hunk 3 above)
Comment (Pierre-Sassoulas) [reply tests/test_similar.py:96]: Haha, good one 😄 We're getting closer and closer to under 700 though !

**Review by Pierre-Sassoulas — COMMENTED** [review]
LGTM !

**Pull request merged**